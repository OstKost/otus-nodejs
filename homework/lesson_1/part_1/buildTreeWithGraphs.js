export const buildTreeWithGraphs = (tree) => {
  return buildLevel(tree);
};

const transformName = (_name) => _name.charAt(0).toUpperCase() + _name.slice(1).toLowerCase();
const renderElement = (count, element) => Array(count).fill(element).join("");

export const buildLevel = (items = [], depth = 0, isLast = false) => {
  let levelResult = '';
  items.forEach((category, index) => {
    const isLastItem = depth ? items.length - 1 === index : false;
    let prefix = depth ? "  " : " ";
    const vLinesLimit = isLast ? 2 : 1;
    const renderCount = (depth - vLinesLimit < 0) ? 0 : (depth - vLinesLimit);
    prefix += depth ? renderElement(renderCount, "│   ") : "";
    if (isLast) prefix += renderElement(1, "    ");
    if (depth) prefix += isLastItem ? "└──" : "├──";

    const name = transformName(category.PRODUCT_TREE_ATTR_NAME);
    const counterChildren = category.children.length ? `(${category.children.length})` : "";
    const title = `${prefix} ${name} ${counterChildren}`.trim();
    levelResult += `${title}\n`;
    levelResult += buildLevel(category.children, depth + 1, isLastItem);
  });
  return levelResult;
};