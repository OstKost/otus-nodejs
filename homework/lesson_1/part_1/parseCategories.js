export const parseCategories = (categories) => {
  const findChildren = (category) => categories.filter(item => item.PRODUCT_TREE_ATTR_PARENT_ID === category.PRODUCT_TREE_ATTR_ID);

  categories.forEach(category => {
    category.children = findChildren(category);
  });

  const topLevel = categories.filter(item => !item.PRODUCT_TREE_ATTR_PARENT_ID);

  return topLevel;
};