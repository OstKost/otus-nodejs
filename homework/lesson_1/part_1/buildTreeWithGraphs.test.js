import { buildLevel, buildTreeWithGraphs } from "./buildTreeWithGraphs.js";
import { parseCategories } from "./parseCategories.js";
import testCategories from "./fixtures/testCategories.json" assert { type: "json" };

describe("Builds correct graphs for tree", () => {
  let graphs;

  beforeAll(() => {
    const tree = parseCategories(testCategories);
    graphs = buildTreeWithGraphs(tree);
  });

  it("Tree renders correctly", () => {
    expect(graphs).toMatchSnapshot();
  });

  it("Top level builds correctly", () => {
    const topLevelGraph = buildLevel(testCategories.slice(0, 2));
    expect(topLevelGraph).toMatchSnapshot();
  });
});