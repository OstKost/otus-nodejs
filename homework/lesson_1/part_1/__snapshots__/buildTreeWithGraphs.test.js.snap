// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builds correct graphs for tree Top level builds correctly 1`] = `
"Категория 1 (3)
├── Категория 1 1 (1)
│   └── Категория 1 1 1
├── Категория 1 2
└── Категория 1 3
Категория 2 (2)
├── Категория 2 1
└── Категория 2 2
"
`;

exports[`Builds correct graphs for tree Tree renders correctly 1`] = `
"Категория 1 (3)
├── Категория 1 1 (1)
│   └── Категория 1 1 1
├── Категория 1 2
└── Категория 1 3
Категория 2 (2)
├── Категория 2 1
└── Категория 2 2
Категория 3 (1)
└── Категория 3 1
Категория 4
"
`;
